pennylane/math/quantum.py

```diff
@@ -720,6 +720,67 @@ def _compute_mutual_info(
     return vn_entropy_1 + vn_entropy_2 - vn_entropy_12
 
 
+def expectation_value(operator_matrix, state_vector, check_state=False, c_dtype="complex128"):
+    r"""Compute the expectation value of an operator and a pure state.
+
+    The expectation value of an operator :math:`A` for a pure states given by state vector :math:`\ket{\psi}`
+    is defined as
+
+    .. math::
+        \langle A \rangle = \bra{\psi} A \ket{\psi}
+
+    Args:
+        operator_matrix (tensor_like): ``(2**N, 2**N)`` or ``(batch_dim, 2**N, 2**N)`` operator matrix.
+        state_vector (tensor_like): ``(2**N)`` or ``(batch_dim, 2**N)`` state vector.
+        check_state (bool): If True, the function will check the validity of both states; that is,
+            the shape and the norm
+        c_dtype (str): Complex floating point precision type.
+
+    Returns:
+        float: Expectation value of the operator for the state vector.
+
+    **Example**
+
+    A operator matrix and a state vectors can be used as arguments and the overlap is returned, e.g.:
+
+    >>> state_vector = np.array([0, 1])
+    >>> operator_matrix = np.array([[0,1],[1,0]])
+    >>> qml.math.expectation_value(operator_matrix, state1)
+    0.0
+
+    .. seealso:: :func:`pennylane.math.fidelity`
+
+    """
+    # Cast as a c_dtype array
+    state_vector = cast(state_vector, dtype=c_dtype)
+
+    # Cannot be cast_like if jit
+    if not is_abstract(state_vector):
+        operator_matrix = cast_like(operator_matrix, state_vector)

```

Comment:
But this also means checking the`dtype` of the inputs, right? 

Also, if the inputs are lists or tuples, that means I could change them maybe using `np.asarray`. What do you think?

Reply:
I see most of the analogous functions, e. g., `qml.math.fidelity`, cast to complex right away, which makes me wonder why 🤔If any of the inputs is compex, `qml.math.einsum` will promote the other or keep both as real. I think these casts may be remnants from before einsum would cast internally. Maybe @eddddddy knows more about it since he changed einsum in #2670?

Reply:
Indeed, this casting is no longer needed in most of the `quantum.py` functions since that PR was merged. We always used the existing functions as a model for new functions that we implement, hence why it appears in almost every function in this file.

If it doesn't cause any issues, then I'd say it's fine to remove it.